File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/clustering/Clustering.java
package cujae.inf.ic.om.heuristic.assignment.clustering;

import java.util.ArrayList;

import cujae.inf.ic.om.controller.tools.Tools;
import cujae.inf.ic.om.heuristic.assignment.Assignment;
import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.output.solution.Cluster;

public abstract class Clustering extends Assignment {
	
	/**
	 * @param  ArrayList<Integer> listado de identificadores
	 * @return ArrayList<Cluster> listado de clusteres
	 * Inicializa los clusters de la solución con los identificadores de listIDElements (clientes y depósitos)
	 **/
	protected ArrayList<Cluster> initializeClusters(ArrayList<Integer> listIDElements){
		ArrayList<Cluster> clusters = new ArrayList<Cluster>();

		Cluster cluster;
		ArrayList<Integer> listIDCustomers;

		int totalElements = listIDElements.size();

		if(totalElements == Problem.getProblem().getDepots().size())
		{
			for(int i = 0; i < totalElements; i++)
			{
				listIDCustomers = new ArrayList<Integer>();
				listIDCustomers.add(listIDElements.get(i).intValue());
				cluster = new Cluster(Problem.getProblem().getDepots().get(i).getIDDepot(), Problem.getProblem().getRequestByIDCustomer(listIDElements.get(i).intValue()), listIDCustomers);
				//cluster = new Cluster(listIDElements.get(i).intValue(), InfoProblem.getProblem().getRequestByIDCustomer(listIDElements.get(i).intValue()), listIDCustomers);
				clusters.add(cluster);
			}
		}
		else // UPGMC
		{
			for(int i = 0; i < totalElements; i++)
			{
				listIDCustomers = new ArrayList<Integer>();

				if(Problem.getProblem().findPosDepot(Problem.getProblem().getDepots(), listIDElements.get(i).intValue()) == -1)
				{
					listIDCustomers.add(listIDElements.get(i).intValue());
					cluster = new Cluster(listIDElements.get(i).intValue(), Problem.getProblem().getRequestByIDCustomer(listIDElements.get(i).intValue()), listIDCustomers);
				}
				else
					cluster = new Cluster(listIDElements.get(i).intValue(), 0.0, listIDCustomers);	// en upgmc no lo agrega al cliente en el cluster

				clusters.add(cluster);	
			}
		}
		
		System.out.println("--------------------------------------------------");
		System.out.println("LISTA DE CLUSTERS");
		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println("ID CLUSTER: " + clusters.get(i).getIDCluster());
			System.out.println("DEMANDA DEL CLUSTER: " + clusters.get(i).getRequestCluster());
			System.out.println("ELEMENTOS DEL CLUSTER: " + clusters.get(i).getItemsOfCluster());
		}
		System.out.println("--------------------------------------------------");

		return clusters;
	}
	
	/**
	 * @param  ArrayList<Customer>
	 * @param  double demanda cubierta del cluster
	 * @param  double capacidad del depósito
	 * @return boolean
	 * Determina si existen clientes que puedan ser asignados al depósito a partir de su demanda
	 */
	protected boolean isFullDepot(ArrayList<Customer> customers, double requestCluster, double capacityDepot){
		boolean isFull = true;

		double currentRequest = capacityDepot - requestCluster;

		if(currentRequest > 0)
		{
			int i = 0;

			while((i < customers.size()) && (isFull))
			{
				if(customers.get(i).getRequestCustomer() <= currentRequest)
					isFull = false;
				else
					i++;
			}
		}	

		return isFull;
		
		// cuando no quedan cliente lo saca como lleno ver si comviene en todos los casos
	}

	protected Location recalculateCentroid(Cluster cluster) {
		double aveAxisX = 0.0;
		double aveAxisY = 0.0;

		int countCustomers = cluster.getItemsOfCluster().size();

		for(int i = 0; i < countCustomers; i++) 
		{	
			Location location = new Location();
			location = Problem.getProblem().getLocationByIDCustomer(cluster.getItemsOfCluster().get(i)); 

			aveAxisX += location.getAxisX();
			aveAxisY += location.getAxisY();
		}

		aveAxisX = (aveAxisX / countCustomers);
		aveAxisY = (aveAxisY / countCustomers);

		Location locationCentroid = new Location();
		locationCentroid.setAxisX(Tools.truncateDouble(aveAxisX, 6));
		locationCentroid.setAxisY(Tools.truncateDouble(aveAxisY, 6));

		return locationCentroid;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/clustering.py
from typing import List
from ..assignment import Assignment
from ...controller.tools.tools import Tools
from ...problem.input.problem import Problem
from ...problem.input.customer import Customer
from ...problem.input.location import Location
from ...problem.solution.cluster import Cluster

class Clustering(Assignment):
    
    def __init__(self):
        super().__init__()
    
    # Método para inicializar una lista de clusters con una lista de elementos.
    def initialize_clusters(self) -> List[Cluster]:
        if not self.list_id_elements:
            raise ValueError("La lista de elementos no puede estar vacía.")
        
        clusters: List[Cluster] = []
        total_elements = len(self.list_id_elements)

        if total_elements == len(Problem.get_problem().get_depots()):
            for i in range(total_elements):
                list_id_customers: List[int] = [self.list_id_elements[i]]
                cluster = Cluster(
                    Problem.get_problem().get_depots()[i].get_id_depot(),
                    Problem.get_problem().get_request_by_id_customer(self.list_id_elements[i]),
                    list_id_customers
                )
                clusters.append(cluster)
        else:
            for i in range(total_elements):
                list_id_customers: List[int] = []
                pos_depot = Problem.get_problem().find_pos_depot(Problem.get_problem().get_depots(), self.list_id_elements[i])
            
                if pos_depot == -1:  # No es un depot
                    list_id_customers.append(self.list_id_elements[i])
                    cluster = Cluster(
                        self.list_id_elements[i], 
                        Problem.get_problem().get_request_by_id_customer(self.list_id_elements[i]), 
                        list_id_customers
                    )
                else:
                    cluster = Cluster(self.list_id_elements[i], 0.0, list_id_customers)
                clusters.append(cluster)

        print("--------------------------------------------------")
        print("LISTA DE CLUSTERS")
        for cluster in clusters:
            print(f"ID CLUSTER: {cluster.get_id_cluster()}")
            print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")
            print(f"ELEMENTOS DEL CLUSTER: {cluster.get_items_of_cluster()}")
        print("--------------------------------------------------")

        return clusters
    
    # Método que determina si existen clientes que puedan ser asignados al depósito a partir de su demanda.
    def is_full_depot(self, customers: List[Customer], request_cluster: float, capacity_depot: float) -> bool:
        is_full: bool = True
        current_request: float = capacity_depot - request_cluster
        if current_request > 0:
            for customer in customers:
                if customer.get_request_customer() <= current_request:
                    is_full = False
                    break
        return is_full
    
    # Método que recalcula el centroide de un cluster basado en las ubicaciones de sus clientes asignados.
    def recalculate_centroid(self, cluster: Cluster) -> Location:
        ave_axis_x = 0.0
        ave_axis_y = 0.0
        count_customers = len(cluster.get_items_of_cluster())

        for customer_id in cluster.get_items_of_cluster():
            location = Problem.get_problem().get_location_by_id_customer(customer_id)
            ave_axis_x += location.get_axis_x()
            ave_axis_y += location.get_axis_y()

        ave_axis_x /= count_customers
        ave_axis_y /= count_customers

        location_centroid = Location()
        location_centroid.set_axis_x(Tools.truncate_double(ave_axis_x, 6))
        location_centroid.set_axis_y(Tools.truncate_double(ave_axis_y, 6))

        return location_centroid